### ISSUE TITLE
Method to free Map resources

### ISSUE THREAD

**Issue by kirtan-desai** [openlayers/openlayers#14517]
**Is your feature request related to a problem? Please describe.**
Currently, there is no method to free up resources other than `map.setTarget(null)` which is not intuitive nor is it mentioned in the documentation. This causes a common bug related to memory leak when you don't destroy the map on component unmount in SPAs.

**Describe the solution you'd like**
Would be nice to have a method like `map.destroy()` or a mention in the documentation that passing null to setTarget frees up the resources.

**Comment by sebakerckhof** [comment]
@ahocevar related to this I have quite some use cases where it would be nice to know when a `Disposable` is disposed. (e.g. cleaning up some related objects etc.)
I think the following additions would make this a lot easier:
- a public `isDisposed()` on `Disposable`
- dispatching a `dispose` event on disposal (`dispose` before `disposeInternal` is called and / or `disposed` after `disposeInternal` is called)

If you'd agree with these additions I'd be happy to make a PR for them.

**Comment by ahocevar** [comment]
@sebakerckhof I don't think this would make much sense, because OpenLayers internally does not call `dispose()` on anything but layer renderers, and `dispose()` is only part of the API for the WebGLTile layer and the Raster source.
